File: src/backend/SunnySeat.Core/Services/ConfidenceCalculator.cs
using NetTopologySuite.Geometries;
using SunnySeat.Core.Entities;

namespace SunnySeat.Core.Services;

/// <summary>
/// Calculates confidence scores for sun exposure calculations based on multiple factors
/// </summary>
public class ConfidenceCalculator
{
    /// <summary>
    /// Calculate comprehensive confidence factors for sun exposure calculation
    /// </summary>
    /// <param name="patio">Patio being analyzed</param>
    /// <param name="shadowInfo">Shadow information for the patio</param>
    /// <param name="solarPosition">Solar position used in calculation</param>
    /// <returns>Detailed confidence factors</returns>
    public ConfidenceFactors CalculateConfidenceFactors(Patio patio,
        PatioShadowInfo shadowInfo, SolarPosition solarPosition)
    {
        var factors = new ConfidenceFactors();

        // Building data quality (40% weight)
        factors.BuildingDataQuality = CalculateBuildingDataQuality(shadowInfo.CastingShadows);

        // Geometry precision (25% weight) 
        factors.GeometryPrecision = CalculateGeometryPrecision(patio);

        // Solar position accuracy (20% weight)
        factors.SolarAccuracy = CalculateSolarAccuracy(solarPosition);

        // Shadow calculation accuracy (15% weight)
        factors.ShadowAccuracy = CalculateShadowAccuracy(shadowInfo, solarPosition);

        // Calculate weighted overall confidence
        factors.OverallConfidence =
            (factors.BuildingDataQuality * 0.40) +
            (factors.GeometryPrecision * 0.25) +
            (factors.SolarAccuracy * 0.20) +
            (factors.ShadowAccuracy * 0.15);

        // Determine confidence category
        factors.ConfidenceCategory = factors.OverallConfidence switch
        {
            >= 0.70 => "High",
            >= 0.40 => "Medium",
            _ => "Low"
        };

        // Identify quality issues and improvement suggestions
        factors.QualityIssues = IdentifyQualityIssues(factors, patio, shadowInfo, solarPosition);
        factors.Improvements = SuggestImprovements(factors, patio, shadowInfo);

        return factors;
    }

    /// <summary>
    /// Calculate comprehensive confidence factors with weather intelligence
    /// </summary>
    /// <param name="patio">Patio being analyzed</param>
    /// <param name="shadowInfo">Shadow information for the patio</param>
    /// <param name="solarPosition">Solar position used in calculation</param>
    /// <param name="weatherData">Weather data for confidence calculation (optional)</param>
    /// <returns>Detailed confidence factors</returns>
    public ConfidenceFactors CalculateConfidenceFactors(Patio patio,
        PatioShadowInfo shadowInfo, SolarPosition solarPosition, WeatherSlice? weatherData)
    {
        var factors = new ConfidenceFactors();

        // Calculate GeometryQuality (60% weight in new formula)
        var geometryQuality = CalculateGeometryQuality(patio, shadowInfo);

        // Calculate CloudCertainty (40% weight in new formula)
        var cloudCertainty = CalculateCloudCertainty(weatherData);

        // New confidence formula: Confidence = (GeometryQuality × 0.6) + (CloudCertainty × 0.4)
        factors.OverallConfidence = (geometryQuality * 0.6) + (cloudCertainty * 0.4);

        // Store component scores for analysis
        factors.GeometryQuality = geometryQuality;
        factors.CloudCertainty = cloudCertainty;

        // Legacy individual factors for backward compatibility
        factors.BuildingDataQuality = CalculateBuildingDataQuality(shadowInfo.CastingShadows);
        factors.GeometryPrecision = CalculateGeometryPrecision(patio);
        factors.SolarAccuracy = CalculateSolarAccuracy(solarPosition);
        factors.ShadowAccuracy = CalculateShadowAccuracy(shadowInfo, solarPosition);

        // Apply confidence caps based on data source and quality
        factors.OverallConfidence = ApplyConfidenceCaps(factors.OverallConfidence, weatherData, patio, shadowInfo);

        // Normalize to percentage (0-1 scale)
        factors.OverallConfidence = Math.Clamp(factors.OverallConfidence, 0.0, 1.0);

        // Determine confidence category
        factors.ConfidenceCategory = factors.OverallConfidence switch
        {
            >= 0.70 => "High",
            >= 0.40 => "Medium",
            _ => "Low"
        };

        // Identify quality issues and improvement suggestions
        factors.QualityIssues = IdentifyQualityIssues(factors, patio, shadowInfo, solarPosition, weatherData);
        factors.Improvements = SuggestImprovements(factors, patio, shadowInfo, weatherData);

        return factors;
    }

    /// <summary>
    /// Calculate GeometryQuality based on building data accuracy (0-1 scale)
    /// </summary>
    private double CalculateGeometryQuality(Patio patio, PatioShadowInfo shadowInfo)
    {
        // Building data completeness
        var buildingDataQuality = CalculateBuildingDataQuality(shadowInfo.CastingShadows);

        // Patio polygon precision
        var polygonPrecision = CalculateGeometryPrecision(patio);

        // Shadow calculation accuracy
        var shadowAccuracy = shadowInfo.Confidence;

        // Weighted average: building 50%, polygon 30%, shadow 20%
        var geometryQuality =
            (buildingDataQuality * 0.5) +
            (polygonPrecision * 0.3) +
            (shadowAccuracy * 0.2);

        return Math.Clamp(geometryQuality, 0.0, 1.0);
    }

    /// <summary>
    /// Calculate CloudCertainty from weather data quality and freshness (0-1 scale)
    /// </summary>
    private double CalculateCloudCertainty(WeatherSlice? weatherData)
    {
        // Missing weather data returns fallback confidence
        if (weatherData == null)
        {
            return 0.5; // 50% base certainty when no weather data
        }

        // Weather data freshness impact
        var dataAge = DateTime.UtcNow - weatherData.CreatedAt;
        var freshnessFactor = CalculateWeatherFreshnessFactor(dataAge);

        // Forecast vs nowcast differentiation
        var forecastFactor = weatherData.IsForecast ? 0.9 : 0.95; // Nowcast gets higher base confidence

        // Source reliability weighting
        var sourceReliability = GetWeatherSourceReliability(weatherData.Source);

        // Combined cloud certainty calculation
        var cloudCertainty = forecastFactor * freshnessFactor * sourceReliability;

        return Math.Clamp(cloudCertainty, 0.0, 1.0);
    }

    /// <summary>
    /// Calculate weather data freshness factor based on age
    /// </summary>
    private double CalculateWeatherFreshnessFactor(TimeSpan dataAge)
    {
        // Fresh data (< 5 min): 100%
        if (dataAge.TotalMinutes < 5)
            return 1.0;

        // Recent data (5-15 min): 95%
        if (dataAge.TotalMinutes < 15)
            return 0.95;

        // Acceptable data (15-30 min): 90%
        if (dataAge.TotalMinutes < 30)
            return 0.90;

        // Aging data (30-60 min): 85%
        if (dataAge.TotalMinutes < 60)
            return 0.85;

        // Old data (1-2 hours): 75%
        if (dataAge.TotalHours < 2)
            return 0.75;

        // Very old data (2-6 hours): 60%
        if (dataAge.TotalHours < 6)
            return 0.60;

        // Stale data (> 6 hours): 40%
        return 0.40;
    }

    /// <summary>
    /// Get weather source reliability weighting
    /// </summary>
    private double GetWeatherSourceReliability(string source)
    {
        return source?.ToLowerInvariant() switch
        {
            "yr.no" => 0.95,           // High reliability - Norwegian Meteorological Institute
            "metno" => 0.95,           // Same as yr.no
            "openweathermap" => 0.85,  // Good reliability
            "openweather" => 0.85,     // Same as openweathermap
            _ => 0.80                  // Default/unknown source
        };
    }

    /// <summary>
    /// Apply confidence caps based on data source and quality
    /// </summary>
    private double ApplyConfidenceCaps(double confidence, WeatherSlice? weatherData,
        Patio patio, PatioShadowInfo shadowInfo)
    {
        var cappedConfidence = confidence;

        // Forecast data: Maximum 90% confidence
        if (weatherData?.IsForecast == true)
        {
            cappedConfidence = Math.Min(cappedConfidence, 0.90);
        }

        // Nowcast data: Maximum 95% confidence
        if (weatherData?.IsForecast == false)
        {
            cappedConfidence = Math.Min(cappedConfidence, 0.95);
        }

        // Missing weather: Maximum 60% confidence
        if (weatherData == null)
        {
            cappedConfidence = Math.Min(cappedConfidence, 0.60);
        }

        // Poor building data: Maximum 70% confidence
        var buildingQuality = CalculateBuildingDataQuality(shadowInfo.CastingShadows);
        if (buildingQuality < 0.6)
        {
            cappedConfidence = Math.Min(cappedConfidence, 0.70);
        }

        return cappedConfidence;
    }

    /// <summary>
    /// Calculate building data quality based on shadow projections
    /// </summary>
    private double CalculateBuildingDataQuality(IEnumerable<ShadowProjection> shadows)
    {
        if (!shadows.Any())
            return 1.0; // No shadows affecting = high confidence

        var qualityScores = shadows.Select(s => s.Confidence);
        return qualityScores.Average();
    }

    /// <summary>
    /// Calculate geometry precision based on patio quality metrics
    /// </summary>
    private double CalculateGeometryPrecision(Patio patio)
    {
        // Use patio quality score from Epic 1
        return patio.PolygonQuality;
    }

    /// <summary>
    /// Calculate solar position accuracy based on sun elevation
    /// </summary>
    private double CalculateSolarAccuracy(SolarPosition solarPosition)
    {
        // Solar position calculations are highly accurate
        // Reduce confidence slightly for very low sun angles
        return solarPosition.Elevation switch
        {
            > 30.0 => 0.98,  // High sun - maximum accuracy
            > 15.0 => 0.95,  // Medium sun - very good accuracy  
            > 5.0 => 0.85,   // Low sun - good accuracy
            > 0.0 => 0.70,   // Very low sun - acceptable accuracy
            _ => 0.50        // Below horizon - limited accuracy
        };
    }

    /// <summary>
    /// Calculate shadow calculation accuracy
    /// </summary>
    private double CalculateShadowAccuracy(PatioShadowInfo shadowInfo, SolarPosition solarPosition)
    {
        // Shadow accuracy depends on sun angle and number of shadow sources
        var baseAccuracy = shadowInfo.Confidence;

        // Reduce confidence for complex shadow scenarios
        var shadowComplexity = shadowInfo.CastingShadows.Count();
        var complexityPenalty = Math.Min(shadowComplexity * 0.03, 0.15); // Max 15% penalty

        // Reduce confidence for very low sun angles (longer, less reliable shadows)
        var elevationPenalty = solarPosition.Elevation < 10.0 ? 0.10 : 0.0;

        return Math.Max(baseAccuracy - complexityPenalty - elevationPenalty, 0.30);
    }

    /// <summary>
    /// Identify quality issues that affect confidence
    /// </summary>
    private List<string> IdentifyQualityIssues(ConfidenceFactors factors, Patio patio,
        PatioShadowInfo shadowInfo, SolarPosition solarPosition, WeatherSlice? weatherData = null)
    {
        var issues = new List<string>();

        if (factors.BuildingDataQuality < 0.7)
            issues.Add("Building height data has low reliability");

        if (factors.GeometryPrecision < 0.7)
            issues.Add("Patio polygon has low quality score");

        if (solarPosition.Elevation < 10.0 && solarPosition.Elevation > 0)
            issues.Add("Sun at low angle - shadow calculations less reliable");

        if (solarPosition.Elevation <= 0)
            issues.Add("Sun below horizon - no direct sunlight");

        if (shadowInfo.CastingShadows.Count() > 5)
            issues.Add("Complex shadow environment with many buildings");

        if (CalculateAreaInSquareMeters(patio.Geometry) < 10.0)
            issues.Add("Very small patio - geometric precision more critical");

        if (factors.OverallConfidence < 0.40)
            issues.Add("Multiple data quality factors reduce overall confidence");

        // Weather-related quality issues
        if (weatherData == null)
        {
            issues.Add("No weather data available - confidence capped at 60%");
        }
        else
        {
            var dataAge = DateTime.UtcNow - weatherData.CreatedAt;
            if (dataAge.TotalHours > 2)
                issues.Add($"Weather data is {dataAge.TotalHours:F1} hours old - reduced confidence");

            if (weatherData.IsForecast)
                issues.Add("Using forecast data - confidence capped at 90%");
        }

        return issues;
    }

    /// <summary>
    /// Suggest improvements to increase confidence
    /// </summary>
    private List<string> SuggestImprovements(ConfidenceFactors factors, Patio patio,
        PatioShadowInfo shadowInfo, WeatherSlice? weatherData = null)
    {
        var improvements = new List<string>();

        if (factors.BuildingDataQuality < 0.7)
        {
            improvements.Add("Survey building heights for more accurate shadow calculations");
            improvements.Add("Verify building data with local planning authorities");
        }

        if (factors.GeometryPrecision < 0.7)
        {
            improvements.Add("Refine patio boundary with higher precision GPS data");
            improvements.Add("Use satellite imagery to improve patio polygon accuracy");
        }

        if (shadowInfo.CastingShadows.Any(s => s.Confidence < 0.7))
        {
            improvements.Add("Update building height data for nearby structures");
        }

        if (factors.OverallConfidence < 0.70)
        {
            improvements.Add("Consider multiple data sources for validation");
            improvements.Add("Use time-averaged calculations to improve reliability");
        }

        // Weather-related improvements
        if (weatherData == null)
        {
            improvements.Add("Integrate weather data for higher confidence scores");
        }
        else
        {
            var dataAge = DateTime.UtcNow - weatherData.CreatedAt;
            if (dataAge.TotalHours > 1)
                improvements.Add("Refresh weather data for improved confidence");
        }

        if (!improvements.Any())
        {
            improvements.Add("Data quality is good - confidence level is appropriate");
        }

        return improvements;
    }

    /// <summary>
    /// Calculate simplified confidence score (0-100) for display purposes
    /// </summary>
    /// <param name="factors">Detailed confidence factors</param>
    /// <returns>Confidence percentage (0-100)</returns>
    public double CalculateDisplayConfidence(ConfidenceFactors factors)
    {
        return Math.Round(factors.OverallConfidence * 100.0, 1);
    }

    /// <summary>
    /// Determine if confidence is sufficient for reliable sun exposure decisions
    /// </summary>
    /// <param name="factors">Confidence factors to evaluate</param>
    /// <returns>True if confidence is sufficient for decision making</returns>
    public bool IsSufficientConfidence(ConfidenceFactors factors)
    {
        return factors.OverallConfidence >= 0.60; // 60% minimum for reliable decisions
    }

    /// <summary>
    /// Calculate area in square meters from geometry (rough approximation)
    /// </summary>
    private double CalculateAreaInSquareMeters(Polygon geometry)
    {
        // Convert from degrees² to m² (rough approximation for Gothenburg latitude ~58°)
        // At this latitude: 1° latitude ≈ 111.3 km, 1° longitude ≈ 55.8 km
        var areaInDegrees = geometry.Area;
        return areaInDegrees * 111300.0 * 55800.0;
    }
}

File: nextjs-app/lib/solar/confidence-calculator.ts
import type {
  SolarPosition,
  VenueShadowInfo,
  WeatherSlice,
  ConfidenceFactors,
  ShadowDataCoverage,
} from './types';
import { getObstructionRiskConfidenceCap } from './obstruction-risk';
import { applyShadowDataCoverageCap } from './shadow-data-coverage';
import { effectiveCloudCover } from './effective-cloud-cover';

/**
 * Two modes:
 * 1. Geometry-only (no weather): 40% building + 25% polygon + 20% solar + 15% shadow
 * 2. Weather-enhanced: 60% GeometryQuality + 40% CloudCertainty, with caps
 */

export function calculateConfidenceFactors(
  polygonQuality: number,
  shadowInfo: VenueShadowInfo,
  solarPosition: SolarPosition,
  weatherData?: WeatherSlice | null
): ConfidenceFactors {
  const buildingDataQuality = calcBuildingDataQuality(shadowInfo);
  const geometryPrecision = polygonQuality;
  const solarAccuracy = calcSolarAccuracy(solarPosition);
  const shadowAccuracy = calcShadowAccuracy(shadowInfo, solarPosition);

  let overallConfidence: number;
  let geometryQuality = 0;
  let cloudCertainty = 0;

  if (weatherData) {
    geometryQuality = calcGeometryQuality(
      buildingDataQuality,
      geometryPrecision,
      shadowInfo.confidence
    );
    cloudCertainty = calcCloudCertainty(weatherData);
    overallConfidence = geometryQuality * 0.6 + cloudCertainty * 0.4;
    overallConfidence = applyConfidenceCaps(
      overallConfidence,
      weatherData,
      buildingDataQuality,
      shadowInfo,
      solarPosition
    );
  } else {
    overallConfidence =
      buildingDataQuality * 0.4 +
      geometryPrecision * 0.25 +
      solarAccuracy * 0.2 +
      shadowAccuracy * 0.15;
    overallConfidence = applyConfidenceCaps(
      overallConfidence,
      null,
      buildingDataQuality,
      shadowInfo,
      solarPosition
    );
  }

  overallConfidence = clamp(overallConfidence, 0, 1);

  const confidenceCategory: ConfidenceFactors['confidenceCategory'] =
    overallConfidence >= 0.7 ? 'High' : overallConfidence >= 0.4 ? 'Medium' : 'Low';

  const qualityIssues = identifyQualityIssues(
    buildingDataQuality,
    geometryPrecision,
    solarPosition,
    shadowInfo,
    overallConfidence,
    weatherData,
    shadowInfo.shadowDataCoverage
  );
  const improvements = suggestImprovements(
    buildingDataQuality,
    geometryPrecision,
    shadowInfo,
    overallConfidence,
    weatherData
  );

  return {
    buildingDataQuality,
    geometryPrecision,
    solarAccuracy,
    shadowAccuracy,
    geometryQuality,
    cloudCertainty,
    overallConfidence,
    confidenceCategory,
    qualityIssues,
    improvements,
  };
}

export function calculateDisplayConfidence(factors: ConfidenceFactors): number {
  return Math.round(factors.overallConfidence * 1000) / 10;
}

export function isSufficientConfidence(factors: ConfidenceFactors): boolean {
  return factors.overallConfidence >= 0.6;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function calcBuildingDataQuality(info: VenueShadowInfo): number {
  if (info.castingShadows.length === 0) {
    if (info.shadowDataCoverage?.allowsHighConfidence === true) return 1.0;
    return applyShadowDataCoverageCap(Math.min(info.confidence, 1.0), info.shadowDataCoverage);
  }
  const average =
    info.castingShadows.reduce((sum, shadow) => sum + shadow.confidence, 0) /
    info.castingShadows.length;
  return applyShadowDataCoverageCap(average, info.shadowDataCoverage);
}

function calcSolarAccuracy(pos: SolarPosition): number {
  if (pos.elevation > 30) return 0.98;
  if (pos.elevation > 15) return 0.95;
  if (pos.elevation > 5) return 0.85;
  if (pos.elevation > 0) return 0.7;
  return 0.5;
}

function calcShadowAccuracy(
  info: VenueShadowInfo,
  pos: SolarPosition
): number {
  const base = info.confidence;
  const complexityPenalty = Math.min(info.castingShadows.length * 0.03, 0.15);
  const elevationPenalty = pos.elevation < 10 ? 0.1 : 0;
  return Math.max(base - complexityPenalty - elevationPenalty, 0.3);
}

function calcGeometryQuality(
  buildingQuality: number,
  polygonPrecision: number,
  shadowConfidence: number
): number {
  return clamp(
    buildingQuality * 0.5 + polygonPrecision * 0.3 + shadowConfidence * 0.2,
    0,
    1
  );
}

function calcCloudCertainty(weather: WeatherSlice): number {
  const ageMs = Date.now() - weather.createdAt.getTime();
  const freshness = weatherFreshnessFactor(ageMs);
  const forecastFactor = weather.isForecast ? 0.9 : 0.95;
  const sourceReliability = weatherSourceReliability(weather.source);
  // STORY 10.1 (AC3, FR12): fold the actual cloud amount into the confidence blend.
  // Until now this function ignored `weather.cloudCover` entirely, so a downpour
  // with fresh Met.no data still scored ~0.9 "cloud certainty" — FR12's promised
  // "weather-based cloud cover uncertainty" was never implemented. The DISPLAYED
  // confidence must fall as cover rises toward total overcast (per the AC), so we
  // multiply the freshness × forecast × source product by a cloud factor that is
  // 1.0 at clear sky and decays linearly to CLOUD_CONFIDENCE_FLOOR at 100% cover.
  // STORY 10.3 (AC2): read the layer-weighted EFFECTIVE cover (via the SAME shared
  // helper the gate uses) rather than the raw total, so confidence and the gate
  // agree on one cloud number — a thin-cirrus sky no longer drops confidence as if
  // it were a blocking deck. UNKNOWN cloud (10.1 AC2 / 10.3 AC3: `effectiveCloudCover`
  // ⇒ `undefined`) stays NEUTRAL — factor 1.0, freshness-only — so a missing-weather
  // slice is NOT penalised as if it were 100% overcast.
  const cloudFactor = cloudConfidenceFactor(effectiveCloudCover(weather));
  return clamp(forecastFactor * freshness * sourceReliability * cloudFactor, 0, 1);
}

/**
 * STORY 10.1 (AC3): map cloud cover (0..100, or `undefined` = unknown) to a
 * confidence multiplier. Clear sky (0%) → 1.0; total overcast (100%) →
 * {@link CLOUD_CONFIDENCE_FLOOR}; unknown cover → 1.0 (neutral, freshness-only).
 * Linear so a re-tune of the floor keeps the monotone "more cloud ⇒ less
 * displayed confidence" property that FR12 requires and the red-first test asserts
 * (relative: 100% materially lower than 0%).
 */
function cloudConfidenceFactor(cloudCover: number | undefined): number {
  if (cloudCover === undefined) return 1;
  const cover = clamp(cloudCover, 0, 100) / 100;
  return CLOUD_CONFIDENCE_FLOOR + (1 - CLOUD_CONFIDENCE_FLOOR) * (1 - cover);
}

// Lowest cloud multiplier, reached at 100% cover. 0.5 makes total overcast cut the
// cloud-certainty term in half — a material, tunable drop (cloud-certainty is 40%
// of the weather-enhanced overall, so a fully-overcast sky pulls overall down by
// up to ~0.2 versus a clear sky with otherwise-identical geometry/freshness).
const CLOUD_CONFIDENCE_FLOOR = 0.5;

function weatherFreshnessFactor(ageMs: number): number {
  const mins = ageMs / 60000;
  if (mins < 5) return 1.0;
  if (mins < 15) return 0.95;
  if (mins < 30) return 0.9;
  if (mins < 60) return 0.85;
  const hrs = mins / 60;
  if (hrs < 2) return 0.75;
  if (hrs < 6) return 0.6;
  return 0.4;
}

function weatherSourceReliability(source: string): number {
  switch (source?.toLowerCase()) {
    case 'yr.no':
    case 'metno':
      return 0.95;
    case 'openweathermap':
    case 'openweather':
      return 0.85;
    default:
      return 0.8;
  }
}

function applyConfidenceCaps(
  confidence: number,
  weather: WeatherSlice | null | undefined,
  buildingQuality: number,
  shadowInfo: VenueShadowInfo,
  solarPosition: SolarPosition
): number {
  let c = confidence;
  if (weather?.isForecast === true) c = Math.min(c, 0.9);
  if (weather?.isForecast === false) c = Math.min(c, 0.95);
  if (!weather) c = Math.min(c, 0.6);
  if (buildingQuality < 0.6) c = Math.min(c, 0.7);
  c = applyShadowDataCoverageCap(c, shadowInfo.shadowDataCoverage);
  c = Math.min(c, getObstructionRiskConfidenceCap(shadowInfo.obstructionRisks));
  if (solarPosition.elevation > 0 && solarPosition.elevation < 5) c = Math.min(c, 0.35);
  else if (solarPosition.elevation > 0 && solarPosition.elevation < 10) c = Math.min(c, 0.55);
  return c;
}

function identifyQualityIssues(
  buildingQuality: number,
  geometryPrecision: number,
  pos: SolarPosition,
  info: VenueShadowInfo,
  overall: number,
  weather?: WeatherSlice | null,
  coverage?: ShadowDataCoverage
): string[] {
  const issues: string[] = [];
  if (buildingQuality < 0.7) issues.push('Building height data has low reliability');
  if (coverage && !coverage.allowsHighConfidence) {
    issues.push('Shadow-caster coverage is not validated for this launch cluster');
  }
  if (geometryPrecision < 0.7) issues.push('Venue polygon has low quality score');
  if (pos.elevation > 0 && pos.elevation < 10) issues.push('Sun at low angle - shadow calculations less reliable');
  if (pos.elevation <= 0) issues.push('Sun below horizon - no direct sunlight');
  if (info.castingShadows.length > 5) issues.push('Complex shadow environment with many buildings');
  if ((info.obstructionRisks?.length ?? 0) > 0) {
    issues.push('Known unmodelled obstruction risk caps confidence');
  }
  if (overall < 0.4) issues.push('Multiple data quality factors reduce overall confidence');
  if (!weather) {
    issues.push('No weather data available - confidence capped at 60%');
  } else {
    const ageHrs = (Date.now() - weather.createdAt.getTime()) / 3600000;
    if (ageHrs > 2) issues.push(`Weather data is ${ageHrs.toFixed(1)} hours old - reduced confidence`);
    if (weather.isForecast) issues.push('Using forecast data - confidence capped at 90%');
  }
  return issues;
}

function suggestImprovements(
  buildingQuality: number,
  geometryPrecision: number,
  info: VenueShadowInfo,
  overall: number,
  weather?: WeatherSlice | null
): string[] {
  const improvements: string[] = [];
  if (buildingQuality < 0.7) {
    improvements.push('Survey building heights for more accurate shadow calculations');
  }
  if (geometryPrecision < 0.7) {
    improvements.push('Refine venue boundary with higher precision GPS data');
  }
  if (info.castingShadows.some((s) => s.confidence < 0.7)) {
    improvements.push('Update building height data for nearby structures');
  }
  if (overall < 0.7) {
    improvements.push('Consider multiple data sources for validation');
  }
  if (!weather) {
    improvements.push('Integrate weather data for higher confidence scores');
  } else {
    const ageHrs = (Date.now() - weather.createdAt.getTime()) / 3600000;
    if (ageHrs > 1) improvements.push('Refresh weather data for improved confidence');
  }
  if (improvements.length === 0) {
    improvements.push('Data quality is good - confidence level is appropriate');
  }
  return improvements;
}

function clamp(v: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, v));
}
